src/index.ts

```diff
@@ -511,7 +511,11 @@ export class BufferedChangeset implements IChangeset {
         ? validationKeys
         : keys(flattenValidations(this.validationMap as object));
 
-    let maybePromise = validationKeys.map(key => this._validateKey(key, this[key]));
+    let maybePromise = validationKeys.map(key => {
+      const value: any = this[key];
+      const resolvedValue = value instanceof ObjectTreeNode ? value.unwrap() : value;
```

Comment:
I think we should never return ObjectTreeNode proxy out of the changeset.
This ensures that validate function won't get ObjectTreeNode. But there is a case which I brought up recently: validating the derived state by defining the getter in the changeset.
This getter will get the values from the BufferedChangeset.get() and it will return ObjectTreeNode proxy at some point.
I think we should move this unwrapping logic to the getter itself and never return ObjectTreeNode from the getter.

Reply:
> unwrapping logic to the getter itself and never return ObjectTreeNode from the getter.

The problem that comes up is we can't reliably provide the whole object, including sibling keys at any level of the tree.  For example (and as you have seen with other PRs as this has developed) is that for a specific node in the tree, it may contain changes, but we also have to merge the `content`.  Otherwise we might be missing sibling keys and fields in a form will all of a sudden null out.

Reply:
Do we need to support such complicated cases when there are nested objects and changes in nested objects? 
I would think that you should create a new Changeset for nested object and put it into the parent changeset if you want to get that behavior.

Reply:
We could send a new Changeset, but that still is a proxy.  `ObjectTreeNode` is just a simpler Proxy. 

This was the previous implementation for reference.

https://github.com/validated-changeset/validated-changeset/blob/43b65252a7d871f1d04e5a46b171f6798935b99f/src/index.ts#L904-L923

Reply:
I meant that we create nested Changeset as a user of that addon.
So, if there is a top-level model and a nested model:
```
class User {
  name: string;
  jobs: Array<Job>;
}

class Job {
  desc: string;
}
```
You create a changeset for top-level object:
```
const user = new User();
const userChangeset = new Changeset(user);
```
If you need to track nested changes in user's jobs you create a nested changeset:
```
const job1 = new Job1();
userChangeset.jobs.push(new Changeset(job1));
```
So, we don't need to guess if we need to merge changes on deep levels as user defined exactly the places where he wants to track nested changes. 

I think it worked in this way when I tried this addon before.

Reply:
Oh I see.  I've never see this way of going about managing nested changes!  

This is what we were aiming to accomplish, just with simpler syntax - `userChangeset.jobs.push(job1);` and with one unified Changeset across the system.  If we can pull it off, then it might be slightly better.  Are we just missing a few spots to unwrap or do you see this approach as fundamentally flawed?

Reply:
I see, better ergonomics is always nice.

I've looked at the code and the tests to understand what the ObjectTreeNode is doing and I realized that we shouldn't unwrap it. I see how it tries to merge original object with changes on any level. 

I've found that it can be kind of hard to use ObjectTreeNode with ember-changeset. When I define a getter on the changeset in order to access relationships I may need to use double get. Example:
```
class Rule extends Model {
  @hasMany user
  @hasMany profile
}
const rule = this.store.createRecord('rule');
const changeset = new Changeset(rule, ..., ...);
Object.defineProperty(changeset, 'isRuleConfigured', {
      get() {
        return this.get('users').get('length') > 0 || this.get('profiles').get('length') > 0;
      },
    });
```
Not sure why, but if I try to use `this.users.length` I don't get proper values.

Generally speaking, I'm using ember-changeset only to manage and validate changes to ED Models, and maybe I wanted something more specific and simple. But maybe I can make it work for my case even with this generic solution.

Reply:
Speaking about unwrapping proxies, I think we need to unwrap Ember.ObjectProxy and Ember.ArrayProxy in the `isObject` util method to support relationship use cases. However, it is not a concern of validated-changeset, it should be overridden in the ember-changeset, however it is hard to override a util method which is not a class method.

Reply:
> how it tries to merge original object with changes on any level.

Just to clarify, there is no merging.  Merging properties was the crux of the problems we had.

> I think we need to unwrap Ember.ObjectProxy and Ember.ArrayProxy
> this.users.length I don't get proper values.

The basic problem is when a nested object in a Changeset is a Proxy (or Ember-specific Proxy).  And the JS language spec doesn't allow us to detect if something is a Proxy (a third party object that is) but we can detect one of our Proxies or an Ember-specific Proxy.  Minor note - could chain 1 get with `users.length` but I still see your point.

Reply:
@andreyfel We can certainly provide our own `#get` trap in ember-changeset that takes into account ObjectProxy and ArrayProxy.

What is `users` in the example you provided?  I'm thinking ObjectProxy isn't the problem with our current setup?  Basically, my feeling is we shouldn't be wrapping an ArrayProxy in ObjectTreeNode proxy.

Reply:
> Minor note - could chain 1 get with get('users.length')

No, it doesn't work if I write it as `this.get('users.length').

> What is users in the example you provided?

`users` is a `hasMany` relationship in the `Rule` model. (I made a typo initially, called it `user`). 

> I'm thinking ObjectProxy isn't the problem with our current setup? 

I think it is still a problem because if ObjectProxy wraps undefined, validated changeset still thinks it is an Object and creates ObjectTreeNode for it.

Reply:
Given the our lack of ability to detect Proxies (they aren't meant to be detected), then I think making changes in ember-changeset to support Ember specifics makes sense.